pallets/msa/src/lib.rs

```diff
@@ -1068,6 +1146,40 @@ impl<T: Config> Pallet<T> {
 		})
 	}
 
+	/// Adds an association between MSA id and ProviderRegistryEntry. As of now, the
+	/// only piece of metadata we are recording is provider name.
+	///
+	/// # Events
+	/// * [`Event::ProviderCreated`]
+	///
+	/// # Errors
+	/// * [`Error::NoKeyExists`] - account does not have an MSA
+	/// * [`Error::ExceedsMaxProviderNameSize`] - Too long of a provider name
+	/// * [`Error::DuplicateProviderRegistryEntry`] - a ProviderRegistryEntry associated with the given MSA id already exists.
+	///
+	pub fn do_create_provider(
+		provider_key: T::AccountId,
+		provider_name: Vec<u8>,
+	) -> DispatchResult {
+		let bounded_name: BoundedVec<u8, T::MaxProviderNameSize> =
+			provider_name.try_into().map_err(|_| Error::<T>::ExceedsMaxProviderNameSize)?;
+
+		let provider_msa_id = Self::ensure_valid_msa_key(&provider_key)?;
+		ProviderToRegistryEntry::<T>::try_mutate(
+			ProviderId(provider_msa_id),
+			|maybe_metadata| -> DispatchResult {
+				ensure!(
+					maybe_metadata.take().is_none(),
+					Error::<T>::DuplicateProviderRegistryEntry
+				);
+				*maybe_metadata = Some(ProviderRegistryEntry { provider_name: bounded_name });
+				Ok(())
+			},
+		)?;
+		Self::deposit_event(Event::ProviderCreated { provider_id: ProviderId(provider_msa_id) });
```

Comment:
I think we should move this out of this function. We have a pattern of logging at the end of extrinsics.

Reply:
It's now on the two extrinsics.